### PR TITLE
Add JAX/EasyDel shared training infrastructure

### PR DESCRIPTION
**Pull request by ndimicTT**
### Issue
N/A (new infrastructure, no associated issue)

### Problem description
JAX/EasyDel experiments currently inline device setup, checkpointing, model loading, and optimizer construction directly in the training script. This makes scripts long, hard to maintain, and inconsistent with Torch experiments that rely on shared helpers in blacksmith/tools/.

### What's Changed
Adds reusable JAX/EasyDel infrastructure modules under blacksmith/tools/, mirroring the existing Torch pattern (DeviceManager, CheckpointManager, etc):

New infra modules (blacksmith/tools/):
- jax_device_manager.py — JAX device selection, TT env setup, mesh creation, data-parallel sharding, and CPU optimizer fallback.
- jax_checkpoint_manager.py — Pickle-based checkpoint save/load with best-N / last-N bookkeeping.
- easydel_helpers.py — EasyDel model loading, LoRA application, and Optax optimizer construction with warmup+cosine decay.
- jax_inspect.py — Prediction display helpers for evaluation.
- workaround_utils_jax.py — GQA attention workaround utilities shared across JAX experiments.

Modified infra modules:
- reproducibility_manager.py — Add JAX seeding branch (framework="jax") and get_jax_rng() helper.
- jax_helpers.py — Add TT-specific clamped softmax CE and masked CE helpers.

Experiment files (from single-chip training branch, unchanged in this PR):

blacksmith/experiments/easydel/qwen/ — Qwen LoRA fine-tuning on SST-2 (training script, configs, data loading, JIT train/eval steps, README, YAML)

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: blacksmith/experiments/easydel/qwen/configs.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    # Model settings
+    model_name: str = Field(default="Qwen/Qwen3-0.6B")
+    max_length: int = Field(default=128, gt=0)
+    dtype: str = Field(default="bfloat16")
+    mask_max_position_embeddings: Optional[int] = Field(default=None)
+
+    @property
+    def jax_dtype(self):
+        key = self.dtype.removeprefix("jnp.")
```
Comment (agobeljicTT) [inline blacksmith/experiments/easydel/qwen/configs.py:29]: self.dtype is already a str without the prefix right? can we just use the return return _DTYPE_MAP[key] and rely on python to fail if there is no entry?

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+from blacksmith.experiments.easydel.qwen.data_loading import (  # noqa: E402
+    load_sst2_batches,
+)
+from blacksmith.experiments.easydel.qwen.train_steps import (  # noqa: E402
+    create_eval_inspect_step_fn,
+    create_eval_step_fn,
+    create_train_step_fn,
+    evaluate,
+)
+from blacksmith.tools.cli import generate_config, parse_cli_options  # noqa: E402
```
Comment (agobeljicTT) [inline blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: do we need these noqa comments

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+):
+    """Invoke the EasyDel Qwen3 causal LM with the kwargs it accepts.
+
+    The call signature is hardcoded on purpose. ``Qwen3ForCausalLM.__call__``
+    is part of EasyDel's stable API for this model family, so there is no
+    need to feature-detect kwargs via ``inspect`` at runtime. If we ever
+    switch to a model whose ``__call__`` takes different kwargs, this
+    function is the single place to update.
+    """
+    return module(input_ids=input_ids, attention_mask=attention_mask)
```
Comment (agobeljicTT) [inline blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: this should not be a separate function as it's too trivial

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
Comment (agobeljicTT) [inline blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: this should be renamed to train.py as per convention (@pglusacTT we should put it in the docs...)

claude's review xd:
```
1. The bundled experiment does not use the new shared tools

  This is the central problem. test_qwen_fine_tuning_easydel.py re-implements almost everything that was just added to blacksmith/tools/:                                                                                                     
   
  ┌────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────┐                                                                                                                     
  │                    Experiment code                     │                   Shared tool equivalent                   │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤                                                                                                                     
  │ _select_preferred_device()                             │ JaxDeviceManager._select_device()                          │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤
  │ load_tokenizer(), load_model()                         │ easydel_helpers.load_tokenizer(), load_easydel_causal_lm() │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤                                                                                                                     
  │ Optimizer construction with warmup+cosine              │ easydel_helpers.build_optimizer()                          │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤                                                                                                                     
  │ LoRA application with CPU context guard                │ easydel_helpers.apply_lora()                               │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤                                                                                                                     
  │ train_steps._clamped_softmax_cross_entropy_per_token() │ jax_helpers.clamped_softmax_cross_entropy_per_token()      │
  ├────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤                                                                                                                     
  │ train_steps._show_predictions()                        │ jax_inspect.show_predictions()                             │
  └────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────┘                                                                                                                     
                            
  The PR description says the experiment files are "unchanged in this PR," implying this is intentional. But that creates a repo state where shared tools exist but nothing uses them. The tools can't be validated against a real workload,  
  and their API may not match actual usage patterns. Either:
  - The experiment should be updated to use the new tools (preferred), or                                                                                                                                                                     
  - The PR should clearly defer the refactor to a follow-up ticket
```

File: blacksmith/experiments/easydel/qwen/train_steps.py [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+    renormalize) to correct the bf16 fused-softmax drift observed on TT.
+
+    Signature::
+
+        eval_step(lora_params, frozen_state,
+                  input_ids, labels, attention_mask) -> loss
+
+    Labels may contain ``-100`` at masked positions.
+    """
+    del device_kind  # dispatch is the same on all devices now
```
Comment (agobeljicTT) [inline blacksmith/experiments/easydel/qwen/train_steps.py]: should we delete this or not provide it in the signature of the function?

Comment (agobeljicTT) [reply blacksmith/experiments/easydel/qwen/train_steps.py]: as well as in inspect_step variant

**Review by agobeljicTT — COMMENTED** [review]
Awesome!
I'd just make a refactor or a copy of the experiment to incorporate the changes so they are testable

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py (hunk 4 above)
Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: I know, the plan for next PR is to fully refactor training script using new infra and to reorganize folder with easydel experiments including names.

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py (hunk 2 above)
Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: Yes, they're needed. The logging.basicConfig(...) call at lines 14-18 is a non-import statement that sits between two groups of imports. Flake8's E402 rule ("module level import not at top of file") fires on every import after that call. Without # noqa: E402, pre-commit would fail. This is the explanation by Claude :)

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py (hunk 3 above)
Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: Agreed and will be changed in the next PR will address easydel folder structure.

File: blacksmith/experiments/easydel/qwen/train_steps.py (hunk 5 above)
Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/train_steps.py]: Done.

File: blacksmith/experiments/easydel/qwen/configs.py (hunk 1 above)
Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/configs.py:29]: Agreed and changed

File: blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py (hunk 3 above)
Comment (agobeljicTT) [reply blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: why not now?

Comment (ndimicTT) [reply blacksmith/experiments/easydel/qwen/test_qwen_fine_tuning_easydel.py]: Done
